nextcloud25: use openssl 1.1 as a PHP extension to fix RC4 encryption

RaitoBezarius opened the pull request:
###### Description of changes

NextCloud encryption is RC4-based, when moving to NextCloud 25 and OpenSSL3, NC cannot decrypt the files anymore, see #197726.

This adds a `openssl-legacy` PHP extension that can be used for these scenarios where OpenSSL 1.1 is easier to use.

Another solution is to provide a custom `openssl.cnf` file to re-enable RC4 encryption to OpenSSL 3, I will open such a PR soon, but it will might be a bit uglier than this one.

cc @mweinelt @Ma27 @AkechiShiro

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

RaitoBezarius committed:
nextcloud25: use openssl 1.1 as a PHP extension to fix RC4 encryption

drupol reviewed (commented):
LGTM but I'll let the seniors approve it.

ajs124 reviewed (approved):
I hate this, but we probably don't really have another choice. Or rather, I don't see how using `openssl_3` and enabling legacy ciphers is better. Let's just hope upstream manages to fix this sometime soon.

What we _could_ do is make this an option and add it to the release notes. That's probably not a better solution either.

I'd love to get some more opinions from some nixos nextcloud and php people.

RaitoBezarius commented:
> I hate this, but we probably don't really have another choice. Or rather, I don't see how using `openssl_3` and enabling legacy ciphers is better. Let's just hope upstream manages to fix this sometime soon.
> 
> What we _could_ do is make this an option and add it to the release notes. That's probably not a better solution either.
> 
> I'd love to get some more opinions from some nixos nextcloud and php people.

We can get this behind a flag to be fair and let people enable it _when they need it_ plus letting people who really want server side encryption complain.

Ma27 reviewed (changes requested):
:-1: from me. I'm pretty sure that this is not the only occurrence in the codebase where PHP's openssl extension is used. If that's the case I don't want to disable OpenSSL 3 for everyone using Nextcloud (just because some people have that security theatre enabled).

I'm fine with merging that, but only if the behavior is hidden behind a flag that's called e.g. `enableBrokenCiphersForSSE` (and yes, the option's name should reflect the fact that you're using broken ciphers IMHO).

RaitoBezarius commented:
> -1 from me. I'm pretty sure that this is not the only occurrence in the codebase where PHP's openssl extension is used. If that's the case I don't want to disable OpenSSL 3 for everyone using Nextcloud (just because some people have that security theatre enabled).
> 
> I'm fine with merging that, but only if the behavior is hidden behind a flag that's called e.g. `enableBrokenCiphersForSSE` (and yes, the option's name should reflect the fact that you're using broken ciphers IMHO).

Totally agreed and asked for some opinions in Matrix chat beforehand regarding this, and I was open to this precise solution. :) (just took the easiest way to not do useless work).

This is now done, I may have been a bit aggressive with the message, let me know your thoughts on them.
Plus, I added a test for this reusing `basic.nix` and I explicitly advise to disable SSE.

RaitoBezarius commented:
@mweinelt do you think I should add more notes in the release notes?

drupol reviewed (commented):
Beautiful!

Question for my own information: Shouldn't `enableBrokenCiphersForSSE` be a global flag in `nixpkgs` ?

mweinelt commented:
> @mweinelt do you think I should add more notes in the release notes?

cc @Mic92 @winterqt

RaitoBezarius commented:
> Beautiful!
> 
> Question for my own information: Shouldn't `enableBrokenCiphersForSSE` be a global flag in `nixpkgs` ?

We enable the OpenSSL 1.x replacement only for this very NixOS module, therefore, it does not prevent anyone from using a NextCloud using OpenSSL 3.x locally, does not prevent anyone from replacing our module with a smarter module for anything I cannot even imagine, etc.

I think very local replacements like this one are ideal, esp. to let advanced users be able to do whatever they want and keep composability.

If a user want to port our mechanism to patch, I provide for a fake OpenSSL PHP Extension called `openssl-legacy` linked against OpenSSL 1.x and can be reused in any NixOS module or PHP environment.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -80,6 +85,35 @@ in {
+        Server-side encryption in NextCloud uses RC4 ciphers, a broken cipher
+        since ~2004.
+
+        This cipher has been disabled in OpenSSL ≥ 3 and require
```
s/require/requires if I'm not mistaken.

RaitoBezarius replied:
done

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -640,6 +674,16 @@ in {
+          See <https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/encryption_configuration.html#disabling-encryption>
+          on how to disable server-side encryption.
+
+          If you think this warning is unwarranted, please open an issue in nixpkgs.
```
IMHO the last sentence can be left out.

RaitoBezarius replied:
IMHO, I see this pattern for warning that can become pestering & annoying and I do not want to give the impression we explored the whole problem space if we made a mistake and anger people, so I think it's cheap to ask for people to come explain for their use case, but that's my 2 cents. (at least, this is what I would wish for.)

Ma27 replied:
The thing is, I'm actually sure you had a different intention (as you stated above), but the wording kind of reminded me of this "X is bad, change my mind" meme. What I want to say, I do think that it people can misunderstand the tone a bit.

But yeah, you do have a point here: how about linking this PR in the warning for more context? If people disagree, they are directly linked to the discussion that lead to this option (to get the full context) and can chime in if they want to.

RaitoBezarius replied:
Agreed!

Ma27 commented:
SGTM, but I'd like to take a closer look when I'm at home (current train has rather slow wifi, so testing is kinda tedious right now).

winterqt commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -80,6 +85,35 @@ in {
+
+    enableBrokenCiphersForSSE = mkOption {
+      type = types.bool;
+      default = false;
```
Maybe we can make this default to true depending on the stateVersion, to avoid breakages? Not sure.

Ma27 replied:
I'm not sure if that's a good idea: IMHO we should only keep openssl_1_1 for Nextcloud whenever people are using SSE. I'd prefer to add a changelog entry informing the user about this breaking change (a brief description is sufficient, it can link to this option for more context, cc @RaitoBezarius )

RaitoBezarius replied:
I see two views:

### Some users are not going to read the changelog

They update, stumble on all their files being broken, rage on NixOS yada yada, someone tell them this is written in changelog, and they say something on : never break user data or whatever even if their data is somewhat still fine except if they do extra manipulations which kill their data (and they will accuse NixOS to not do the right thing™ — whatever).

### Some users read the changelog

They see that we set OpenSSL 1.x as default for SSE backward compatibility people, they are not interested into that, and they disable it right away during upgrade.
It is an extra manual step during their upgrade, but it's fine ; worst case, they blame NixOS to not be sufficiently proactive on security posture : well, it is always an equilibrium between broken and secure.

---

Personally, I prefer forcing NC on 3.x for the next release and I do not think it is hard to fix it even if you discover that all your files are non-decrypting on your upgrade. Stumbling and looking for Internet yield answer quite quickly.
Though, it requires you to notice the journal logs.

At the same time, I see the scenario of the first heading happening so easily, and we may have people with corrupted data because they tried to force decryption or did something strange, and, yes, people do not read nixpkgs warnings (I remember someone over social media complaining about something even if the warning was ~explicit.)

I don't think there is a good choice anyway, given what I exposed above. If someone has a strong feeling about this and is followed by enough contributors, we can implement that?

RaitoBezarius replied:
Nevertheless, I implemented the changelog entry as a backward incompatibility for now.

RaitoBezarius replied:
I added the suggested workaround as a separate commit, if we do take the decision to keep it, I squash, otherwise, I drop.

winterqt reviewed (commented):
> @mweinelt do you think I should add more notes in the release notes?

I'm not sure right now, see below.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -640,6 +674,16 @@ in {
         ++ (optional (versionOlder cfg.package.version "25") (upgradeWarning 24 "22.11"))
+        ++ (optional cfg.enableBrokenCiphersForSSE ''
+          You enabled OpenSSL 1.x for NextCloud, this is not necessary except if you
+          are running a RC4-based server side encryption, we advise you to disable server-side
```
Hmm... with respect to the discussion below I'm not sure if that's a bit too harsh. How about something like "You're using PHP's openssl extension built against OpenSSL 1.1. This is only necessary if you're using server-side encryption. Please keep in mind that it's using the broken RC4 cipher"?

RaitoBezarius replied:
That's a lot better than my version :) thank you!

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -13,7 +13,12 @@ let
       (with all;
-        enabled
+        # disable default openssl extension
+        (lib.filter (e: e != cfg.phpPackage.extensions.openssl) enabled)
```
Can't we filter out by the package's pname for instance? This would also make sure that overlayed `php.extensions.openssl` is correctly filtered out.

RaitoBezarius replied:
We definitely can

RaitoBezarius commented:
Mentioning https://github.com/nextcloud/server/pull/25551 for upstream tracking.

Ma27 commented on nixos/doc/manual/release-notes/rl-2211.section.md:
```diff
@@ -181,6 +181,8 @@ Available as [services.patroni](options.html#opt-services.patroni.enable).
 
 - The `p4` package now only includes the open-source Perforce Helix Core command-line client and APIs. It no longer installs the unfree Helix Core Server binaries `p4d`, `p4broker`, and `p4p`. To install the Helix Core Server binaries, use the `p4d` package instead.
 
+- The NextCloud NixOS module uses OpenSSL 3.x for its PHP's openssl extension, this breaks RC4-based server-side encryption in NextCloud, making all your files unreadable upon upgrade, no data loss **should occur** though. To restore functionality, `services.nextcloud.enableBrokenCiphersForSSE` has to be set to `true`. NextCloud is planning to implement AES-256-GCM server-side encryption in the f ...
```
Please link the option via `[](#opt-services.nextcloud.enableBrokenCiphersForSSE)`. Then people can directly navigate to the option's decription for further context.

Also, I'm wondering if we should check whether that's actually the case (I can do that this week).

Ma27 replied:
> I can do that this week

Just did it:

* with openssl3 for php installed, you can neither download nor upload files (as expected).
* you can disable encryption and upload new files (old files still inaccessible)
* you can install openssl1_1 and re-enable encryption and force decrypt all

IoW we can write something like "upon testing we couldn't trigger any cases of data-loss, but we cannot guarantee that for every accidental openssl upgrade".

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -80,6 +85,35 @@ in {
+        your files anymore, enabling this option can restore access to your files.
+
+        It is advised to [disable server-side encryption](https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/encryption_configuration.html#disabling-encryption) as it is unclear
+        it provides any amount of security beyond encryption for S3-style buckets
```
I'd probably say "external storage" rather than S3 as the reasoning applies to any kind of storage you're not operating on your own. Also, I'd probably alter the wording a bit here: "Unless you're using external storage, it's advised to [disable server-side encryption](...)`.

RaitoBezarius replied:
Done.

Ma27 reviewed (approved):
Apart from two nitpicks and the default for `enableBrokenCiphersForSSE` it looks fine, so I'll give my (tentative) approval (please don't merge yet, we should resolve these things first!).

I'd also test whether I can trigger any data corruption with openssl3 (by simulating an upgrade from a Nextcloud on 22.05 to master with SSE enabled), then we have at least **some** datapoints.

Finally, big thanks @RaitoBezarius for tackling this and your patience throughout all the feedback cycles :)

RaitoBezarius commented:
> My 5 cents: As long as OpenSSL 1.1 is still supported (until 2023/09) we could go back for 22.11 and wait for this mess to resolve itself in that time and ideally not make this a hassle for anyone.

Do I take this as: you prefer to enable by default the workaround?

RaitoBezarius committed:
nextcloud25: enable by default broken ciphers for NixOS ≤ 22.11

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -80,6 +85,37 @@ in {
+
+    enableBrokenCiphersForSSE = mkOption {
+      type = types.bool;
+      # Workaround can be removed at backport-time for 22.11.
```
What does that mean?

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -80,6 +85,37 @@ in {
+    enableBrokenCiphersForSSE = mkOption {
+      type = types.bool;
+      # Workaround can be removed at backport-time for 22.11.
+      default = !(versionOlder stateVersion "22.11");
```
Please add `defaultText` here: otherwise building the manual depends on actual system config (even though it's not too much of an issue since stateVersion rarely changes).

Ma27 committed:
nixos/nextcloud: minor docs cleanup for openssl change

* s/NextCloud/Nextcloud/g
* `enableBrokenCiphersForSSE` should be enabled by default for any NixOS
  installation from before 22.11 to make sure existing installations
  don't run into the issue. Not the other way round.
* Update release notes to reflect on that.
* Improve wording of the warning a bit: explain which option to change
  to get rid of it.
* Ensure that basic tests w/o `enableBrokenCiphersForSSE` run with
  OpenSSL 3.

Ma27 commented:
Pushed a few minor docs fixes, the change is fine by me now.

In case noone else finds something, I'd be fine with merging that.

RaitoBezarius commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -88,32 +88,32 @@ in {
       type = types.bool;
-      # Workaround can be removed at backport-time for 22.11.
-      default = !(versionOlder stateVersion "22.11");
+      default = versionOlder stateVersion "22.11";
```
Hm, shouldn't that be a not here?
I read it as: `if stateVersion > 22.11 then true else false`, meaning, broken ciphers for SSE are enabled for ≥ 23.05 and disabled for ≤ 22.11, which is the contrary of what we want. Did I miss something about the logic of that function?

drupol replied:
I read the opposite: `if stateVersion < 22.11 then true else false`

Older version means smaller version number, right ?

RaitoBezarius replied:
Worked too much, today, I guess :D — ignore my comment please

Ma27 replied:
@drupol I guess the confusion stems from the ordering of the arguments (and I can totally related to that, I always have to think twice as well, otherwise I would've spotted it yesterday already ;-) )

`versionOlder` evaluates to `true` if the first argument is smaller (i.e. older) than the second one:

```
nix-repl> :l <nixpkgs/lib>
Added 402 variables.

nix-repl> stateVersion = "22.05"

nix-repl> versionOlder stateVersion "22.11"
true
```

And this is precisely what we want.

Ma27 commented:
OK, found another small thing to be fixed before merging, please don't do that right now.

I guess to be safe, I should implement a VM test for that.

Ma27 committed:
nixos/nextcloud: fixup openssl compat change

Upon testing the change itself I realized that it doesn't build properly
because

* the `pname` of a php extension is `php-<name>`, not `<name>`.
* calling the extension `openssl-legacy` resulted in PHP trying to compile
  `ext/openssl-legacy` which broke since it doesn't exist:

      source root is php-8.1.12
      setting SOURCE_DATE_EPOCH to timestamp 1666719000 of file php-8.1.12/win32/wsyslog.c
      patching sources
      cdToExtensionRootPhase
      /nix/store/48mnkga4kh84xyiqwzx8v7iv090i7z66-stdenv-linux/setup: line 1399: cd: ext/openssl-legacy: No such file or directory

I didn't encounter that one before because I was mostly interested in
having a sane behavior for everyone not using this "feature" and the
documentation around this. My findings about the behavior with turning
openssl1.1 on/off are still valid because I tested this on `master` with
manually replacing `openssl` by `openssl_1_1` in `php-packages.nix`.

To work around the issue I had to slightly modify the extension
build-system for PHP:

* The attribute `extensionName` is now relevant to determine the output
  paths (e.g. `lib/openssl.so`). This is not a behavioral change for
  existing extensions because then `extensionName==name`.

  However when specifying `extName` in `php-packages.nix` this value is
  overridden and it is made sure that the extension called `extName` NOT
  `name` (i.e. `openssl` vs `openssl-legacy`) is built and installed.

  The `name` still has to be kept to keep the legacy openssl available
  as `php.extensions.openssl-legacy`.

Additionally I implemented a small VM test to check the behavior with
server-side encryption:

* For `stateVersion` below 22.11, OpenSSL 1.1 is used (in `basic.nix`
  it's checked that OpenSSL 3 is used). With that the "default"
  behavior of the module is checked.

* It is ensured that the PHP interpreter for Nextcloud's php-fpm
  actually loads the correct openssl extension.

* It is tested that (encrypted) files remain usable when (temporarily)
  installing OpenSSL3 (of course then they're not decryptable, but on a
  rollback that should still be possible).

Finally, a few more documentation changes:

* I also mentioned the issue in `nextcloud.xml` to make sure the issue
  is at least mentioned in the manual section about Nextcloud. Not too
  much detail here, but the relevant option `enableBrokenCiphersForSSE`
  is referenced.

* I fixed a few minor wording issues to also give the full context
  (we're talking about Nextcloud; we're talking about the PHP extension
  **only**; please check if you really need this even though it's
  enabled by default).

  This is because I felt that sometimes it might be hard to understand
  what's going on when e.g. an eval-warning appears without telling where
  exactly it comes from.

Ma27 commented:
Rebased onto `master` and pushed another fix (see commit message for all the details).

Considering that this now touches PHP stuff and I contributed quite a bunch of stuff in here, I'd like to have more reviewers.
cc @ajs124 @drupol @RaitoBezarius @yayayayaka

ajs124 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -80,6 +85,40 @@ in {
+      description = lib.mdDoc ''
+        This option enables using the OpenSSL PHP extension linked against OpenSSL 1.1
+        rather than latest OpenSSL (≥ 3), this is not recommended unless you need
+        it for server-side encryption (SSE). SSE uses the legacy RC4 cipher which is
+        considered broken for several years now.
```
If you want, you can mention [RFC7465](https://datatracker.ietf.org/doc/html/rfc7465) for RC4 deprecation.

ajs124 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -649,6 +688,23 @@ in {
+
+          If you need to use server-side encryption you can ignore this waring.
+          Otherwise you'd have to disable server-side encryption first in order
+          to be able to safely disable this option and get rid of that warning.
```
`this` or `that` warning?

Ma27 commented:
@ajs124 fixed :)

etu reviewed (approved):
As for the changes to the php packages I think this looks fine

Ma27 commented:
cc @NixOS/release-engineers @NixOS/nixos-release-managers this should really land in 22.11. Would be cool to get a few more reviewers here first, though.

cab404 reviewed (approved):
Wow, that's some thorough testing.

AkechiShiro commented:
A fix for the root issue is coming up in Nextcloud 26 : https://github.com/nextcloud/server/issues/32003#issuecomment-1443547680

Tagging maintainers of Nextcloud so that they be notified about this coming change : 
@schneefux @bachp @globin @fpletz

I guess this fix needs a tweak so that it can only be used on `pkgs.nextcloud25` and lower, once `pkgs.nextcloud` defaults to Nextcloud 26.

Should I be trying my hand at making a PR to adapt the Nextcloud module to  take into account that fix should be allowed to **run only** on NC25 and lower versions. And ignore the fix for NC26 by using the function `versionOlder` and `cfg.package.version` ? 

I'd guess some testing would be needed to make sure nothing breaks down on NC26, there are only beta version of the server, so testing would wait and will be possible when NC26 stable or release candidate is out.

tl;dr :
Nextcloud 26 will not depend on OpenSSL for RC4 server side encryption (AFAIK downgrading OpenSSL will not be necessary anymore for server side encryption to be working), see these pending PR for Nextcloud 26 for more information (both PRs were merged in Nextcloud 26 beta 5) : https://github.com/nextcloud/server/pull/36173#issuecomment-1441380510 https://github.com/nextcloud/server/pull/35916

Ma27 commented:
> A fix for the root issue is coming up in Nextcloud 26

I wouldn't call it a fix for the root issue though, rather working around the root cause ;-)

> Should I be trying my hand at making a PR to adapt the Nextcloud module to take into account that fix should be allowed to run only on NC25 and lower versions

We don't have nc26 yet, so I guess, no action needed here until the release is out. Thanks for the heads-up, I'll adapt the module then accordingly.

Of course, if you're faster when it's out, I won't stop you from filing a PR.

chayleaf commented:
nc26 is released now!

RaitoBezarius commented:
We already adapted NC26 IIRC. We still need our workaround for NC25 as long as it is a supported version.

chayleaf commented:
I think it still uses openssl-legacy with NC26, at least I couldn't find anything for that in https://github.com/NixOS/nixpkgs/blob/2fdee63799a6cb2cb074ac62c09665b377d261ea/nixos/modules/services/web-apps/nextcloud.nix#L21

nh2 commented:
I want to upgrade from NixOS 22.11 to 23.05 with Nextcloud installed.

I get the output:

```
error: Package ‘openssl-1.1.1u’ in /nix/store/51p9r89zh93piqkmr18h4m7mkf10a1sg-nixos-23.05/nixos/pkgs/development/libraries/openssl/default.nix:210 is marked as insecure, refusing to evaluate.


       Known issues:
        - OpenSSL 1.1 is reaching its end of life on 2023/09/11 and cannot be supported through the NixOS 23.05 release cycle. https://www.openssl.org/blog/blog/2023/03/28/1.1.1-EOL/
```

Unfortunately there is no mention at all of Nextcloud, even though it seems to be the problem.

I only found it with an invocation from https://github.com/NixOS/nixpkgs/issues/210452#issuecomment-1616777726

```sh
cd /etc/nixos
NIXPKGS_ALLOW_INSECURE=1 nix-instantiate --strict "<nixpkgs/nixos>" -A system
```

which prints:

```
trace: warning: You're using PHP's openssl extension built against OpenSSL 1.1 for Nextcloud.
This is only necessary if you're using Nextcloud's server-side encryption.
Please keep in mind that it's using the broken RC4 cipher.

If you don't use that feature, you can switch to OpenSSL 3 and get
rid of this warning by declaring

  services.nextcloud.enableBrokenCiphersForSSE = false;

If you need to use server-side encryption you can ignore this warning.
Otherwise you'd have to disable server-side encryption first in order
to be able to safely disable this option and get rid of this warning.
See <https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/encryption_configuration.html#disabling-encryption> on how to achieve this.

For more context, here is the implementing pull request: https://github.com/NixOS/nixpkgs/pull/198470
```

It would be good if `nixos-rebuild boot` could be convinced to output a better error message / produce this warning, vs leaving the user in the dark on _which_ package depends on old OpenSSL.

Ma27 commented:
This isn't a problem with Nextcloud's packaging, but with how warnings currently work in nixpkgs (the problem appears with every dependant of openssl 1.1.1 on 23.05 and later), this might be subject to change with https://github.com/NixOS/rfcs/pull/127/files though. In other words, this is IMHO the wrong place to report the issue because it's not a problem with this package and this is effectively necro-bumping an old PR (this wasn't a problem when this change was merged).

In other words, please file a new issue for that :)
